streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -1081,8 +1081,9 @@ private void shutdown(final boolean cleanRun) {
         streamsMetrics.removeAllSensors();
     }
 
+    // visible for testing
     @SuppressWarnings("ThrowableNotThrown")
-    private void shutdownTasksAndState(final boolean cleanRun) {
+    void shutdownTasksAndState(final boolean cleanRun) {
```

Comment:
Personally I do not like making an internal function visible only because it is needed in unit tests, and as long as such access is rare (and should be) I'm fine with reflections actually; I agree we do not have such rules universally. So let me propose a different way just for this PR (we can consider what's the best approach globally): let the thread just call `close` in the unit test with mock producer / consumer clients, instead of this function. After all the unit tests are for class's public API's behaviors. WDYT @dguy @mjsax ?